### ISSUE TITLE
Segmentation fault (core) with 3.10.0 on strtr() on some chars if passed array exeeds 64 items

### ISSUE THREAD

**Issue by keepxtreme** [facebook/hhvm#6384]
Hi everyone,

starting with the update 3.9.1 to 3.10.0 dokuwiki starts crashing on Mainpage (doku.php). The hhvm/error.log only reports segmentation fault: core dumped and refers to a tmp-log:

PHP Stacktrace:
#0  utf8_deaccent(:, -1) called at [/var/www/service/wiki/inc/pageutils.php:132]
#1  cleanID(:) called at [/var/www/service/wiki/inc/pageutils.php:294]
#2  wikiFN(:, ) called at [/var/www/service/wiki/inc/pageutils.php:549]
#3  resolve_pageid(, :, ) called at [/var/www/service/wiki/inc/template.php:941]
#4  tpl_youarehere() called at [/var/www/service/wiki/lib/tpl/dokuwiki/tpl_header.php:84]
#5  include(/var/www/service/wiki/lib/tpl/dokuwiki/tpl_header.php) called at [/var/www/service/wiki/lib/tpl/dokuwiki/main.php:33]
#6  include(/var/www/service/wiki/lib/tpl/dokuwiki/main.php) called at [/var/www/service/wiki/inc/actions.php:196]
#7  act_dispatch() called at [/var/www/service/wiki/doku.php:124]

I tested the function utf8_deaccent and it crashes if only a ":" string is passed using the specified array:

````PHP
if(!function_exists('utf8_deaccent')){
    /**
     * Replace accented UTF-8 characters by unaccented ASCII-7 equivalents
     *
     * Use the optional parameter to just deaccent lower ($case = -1) or upper ($case = 1)
     * letters. Default is to deaccent both cases ($case = 0)
     *
     * @author Andreas Gohr <andi@splitbrain.org>
     *
     * @param string $string
     * @param int $case
     * @return string
     */
    function utf8_deaccent($string,$case=0){
        if($case <= 0){
            global $UTF8_LOWER_ACCENTS;
            $string = strtr($string,$UTF8_LOWER_ACCENTS);
        }
        if($case >= 0){
            global $UTF8_UPPER_ACCENTS;
            $string = strtr($string,$UTF8_UPPER_ACCENTS);
        }
        return $string;
    }
}

**Comment by keepxtreme** [comment]
ok, I digged a lot deeper and it seems like strst() fails if the passed array exeeds 64 pairs

**Comment by matzekuh** [comment]
I tried to reproduce the problem on 3v4l. I issued `strt("i",$UTF8_UPPER_ACCENTS)`.

Case 1: `$UTF8_UPPER_ACCENTS` with 64 entries not containing `":"`.
https://3v4l.org/BnTDC works fine.

Case 2: `$UTF8_UPPER_ACCENTS` with 65 entries not containing `":"`.
https://3v4l.org/HMLUV hhvm-3.10.0 exited with code 137

Case 3: `$UTF8_UPPER_ACCENTS` with 65 entries containing `":"`.
https://3v4l.org/prZbb works fine.

For every case the example works in hhvm-3.9.1.

**Comment by jwatzman** [comment]
Interesting -- sounds like you found a repro case on 3v4l, which is really useful. Someone will take a look at this when we get a chance, thanks!

**Comment by keepxtreme** [comment]
3.10.0 crashs in Drupals Admin Menu module in the function admin_menu_tree(): http://www.drupalcontrib.org/api/drupal/contributions%21admin_menu%21admin_menu.inc/function/admin_menu_tree/7
This is as well related to strtr() as commenting out the line containing the function brings the function back to work. However, as this module is enabled on a whole bunch of drupal sites I usume it as critical to get this fixed

**Comment by SiebelsTim** [comment]
I did not do a bisect, but I'm fairly certain that https://github.com/facebook/hhvm/commit/b94ac59fcc89d766c29551d02eeab8e84d3d5f43 is the commit that introduced the issue.

My reasoning is, that commit introduced a new path for `strtr` when the array has more than 64 elements (kBitsPerQword == 64).

Furthermore I know that [WuManberReplacement](https://github.com/facebook/hhvm/blob/4a3be43df46ea29b5d7860aaa524bd3ee8aaf4c8/hphp/runtime/ext/string/ext_string.cpp#L1765) is initialized with a `minLen` of 2 which results in a unsigned overflow [here](https://github.com/facebook/hhvm/blob/4a3be43df46ea29b5d7860aaa524bd3ee8aaf4c8/hphp/runtime/ext/string/ext_string.cpp#L1890), because your source string is one character long. 
You will see no segfault when you increase your source string length. 

I've got to run so I can't look at this right now. Maybe later today or in a few days.

**Comment by jwatzman** [comment]
@SiebelsTim awesome, thanks for the investigation! Does seem likely, I'll confirm later if you haven't gotten around to it yet. I've poked the author on the original PR to take a look here; if they don't have a satisfactory fix shortly, we can revert this.

**Comment by SiebelsTim** [comment]
Fix is up for review and I could confirm that it segfault when every element of the array contains a string larger than the source.
https://3v4l.org/f0LPD

**Comment by jwatzman** [comment]
You are awesome! I'll suck that in for review shortly.

**Comment by ebernhardson** [comment]
This is fixed in a more comprehensive patch here:  https://github.com/ebernhardson/hiphop-php/commit/8b9d87033236af15801e1e92a752a99adc31640a

I'll try and figure out what happened with that patch, its not turning up in github search for hhvm, while i'm certain i submitted it as a PR

**Comment by SiebelsTim** [comment]
@ebernhardson https://github.com/facebook/hhvm/pull/6071 this one?

**Comment by ebernhardson** [comment]
Yup thats the one, I'm not sure what went wrong with that patch though?

I submitted an updated version that fixes all segfaults we see in production here on Aug 19, at the end of september it looks to have been closed with a human timeout, but i'm not sure what about the patch wasn't sufficient to merge.

**Comment by SiebelsTim** [comment]
It contained non-ascii characters in a test. See https://reviews.facebook.net/D44973

**Comment by ebernhardson** [comment]
ahh, ok i can fix that up relatively easily.

**Comment by jwatzman** [comment]
I'll reopen that PR. Should we wait on it, or take https://github.com/facebook/hhvm/pull/6395 as a stopgap in the meantime?

**Comment by SiebelsTim** [comment]
@jwatzman Looks like the only thing stopping his PR right now is a malformed test. As long as he gets to it, take his PR.

**Comment by ebernhardson** [comment]
I've just pushed  a new version without non-ascii characters, they wern't necessary that set of patterns is just how we happened to notice it in our own production environment.

**Comment by jwatzman** [comment]
Awesome, thanks! I've pulled in the update for a test run, will be a few hours. I'll comment further on the Phabricator diff if anything else needs to happen.

**Comment by jwatzman** [comment]
https://github.com/facebook/hhvm/commit/7cc39aa57af8b8e20a71c0e0291369a8f805efcf should fix this issue -- tomorrow, can you run a nightly build to confirm? (Or build from source with that rev to be sure.) I'm going to merge into 3.10 now.

**Comment by keepxtreme** [comment]
@jwatzman positive, nightly 21-10-2015 solves the issue for us, thanks. However, I would appreciate having this patch in the stable release...

**Comment by SiebelsTim** [comment]
@keepxtreme 
>  I'm going to merge into 3.10 now.

;-)

**Comment by jwatzman** [comment]
Yeah it's in the branch and will be out whenever we do 3.10.1. Discussing now how when we want to do that.

**Comment by jwatzman** [comment]
I'm going to try to roll 3.10.1 with this later today.

**Comment by gmagana** [comment]
@jwatzman Is there a changelog or release notes for 3.10 to see what's included in the release? I relied in the past on the blog (http://hhvm.com/blog/), but there is no mention of 3.10, wondering if there is a better place to look.

**Comment by jwatzman** [comment]
We haven't published a blog post, largely because no one has done the legwork of collating the major changes in 3.10 :)

**Comment by jwatzman** [comment]
Blog post is up, and 3.10.1 is building right now. Going to close this out. Thanks for your patience everyone!

**Comment by gmagana** [comment]
Thank you so much!
